rustworkx-core/src/coloring.rs

```diff
@@ -220,8 +221,218 @@ where
+    // Partially assigned colors (indexed by internal edge index)
+    colors: Vec<Option<usize>>,
+}
+
+impl<G> MisraGries<G>
+where
+    G: EdgeIndexable + IntoEdges,
+{
+    pub fn new(graph: G) -> Self {
+        let colors = vec![None; graph.edge_bound()];
+        MisraGries { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: G::NodeId) -> HashSet<usize> {
+        let used_colors: HashSet<usize> = self
+            .graph
+            .edges(u)
+            .filter_map(|edge| self.colors[EdgeIndexable::to_index(&self.graph, edge.id())])
+            .collect();
+        used_colors
+    }
+
+    // Returns the smallest free (aka unused) color at node u
+    fn get_free_color(&self, u: G::NodeId) -> usize {
+        let used_colors = self.get_used_colors(u);
+        let free_color: usize = (0..)
+            .position(|color| !used_colors.contains(&color))
+            .unwrap();
+        free_color
+    }
+
+    // Returns true iff color c is free at node u
+    fn is_free_color(&self, u: G::NodeId, c: usize) -> bool {
+        let used_colors = self.get_used_colors(u);
+        !used_colors.contains(&c)
+    }
+
+    // Returns the maximal fan on edge (u, v) at u
+    fn get_maximal_fan(
+        &self,
+        eid: G::EdgeId,
+        u: G::NodeId,
+        v: G::NodeId,
+    ) -> Vec<(G::EdgeId, G::NodeId)> {
+        let mut fan: Vec<(G::EdgeId, G::NodeId)> = Vec::new();
+        fan.push((eid, v));
+
+        let mut neighbors: Vec<(G::EdgeId, G::NodeId)> = self
+            .graph
+            .edges(u)
+            .map(|edge| (edge.id(), edge.target()))
+            .collect();
+
+        let mut last_node_in_fan = v;
+        neighbors.remove(neighbors.iter().position(|x| x.1 == v).unwrap());
+
+        let mut fan_extended: bool = true;
+        while fan_extended {
+            fan_extended = false;
+
+            for (edge, node) in &neighbors {
+                if let Some(color) = self.colors[EdgeIndexable::to_index(&self.graph, *edge)] {
+                    if self.is_free_color(last_node_in_fan, color) {
+                        fan.push((*edge, *node));
+                        last_node_in_fan = *node;
+                        fan_extended = true;
+                        neighbors.remove(neighbors.iter().position(|x| x.1 == *node).unwrap());
+                        break;
+                    }
+                }
+            }
+        }
+
+        fan
+    }
+
+    // Assuming that color is either c or d, returns the other color
+    fn flip_color(&self, c: usize, d: usize, e: usize) -> usize {
+        if e == c {
+            d
+        } else {
+            c
+        }
+    }
+
+    // Returns the longest path starting at node u with alternating colors c, d, c, d, c, etc.
+    fn get_cdu_path(&self, u: G::NodeId, c: usize, d: usize) -> Vec<(G::EdgeId, usize)> {
+        let mut path: Vec<(G::EdgeId, usize)> = Vec::new();
+        let mut cur_node: G::NodeId = u;
```

Comment:
The compiler should be able to infer the type here because it's defined on `u`.

```suggestion
        let mut cur_node = u;
```

Reply:
Removed in [7e6470d](https://github.com/Qiskit/rustworkx/pull/902/commits/7e6470dc628dc7576f52666e71ce313e362b7e77).